### ISSUE TITLE
UselessTernaryOperator false positive with array elements

### ISSUE THREAD

**Issue by tominventisbe** [slevomat/coding-standard#655]
The `UselessTernaryOperator` sniff claims the following ternary operator is redundant:

```php
$array = ['test' => foo()];
$bar = $array['test'] ? true : false;
```

It actually isn't redundant in this case, since without it the value of the array element would be taken over, which is not necessarily a boolean itself.

A workaround seems to be to use `!!$array['test']` instead of the ternary operator - a cast may do the same trick.

**Comment by kukulich** [comment]
@tominventisbe What version do you use?

**Comment by tominventisbe** [comment]
@kukulich Thanks for your response. The very latest at the moment, version 5.0.4.

**Comment by kukulich** [comment]
@tominventisbe Did you read README for this sniff?

**Comment by kukulich** [comment]
The reported error is right, the ternary operator is useless. So you should rewrite your condition.

**Comment by tominventisbe** [comment]
@kukulich Thanks for your response. I did take a look at the `assumeAllConditionExpressionsAreAlreadyBoolean` option, but I don't believe it's related to this problem (?). I also took a look inside the tests, in case I just understood the purpose incorrectly.

`assumeAllConditionExpressionsAreAlreadyBoolean` value is also set to `false` by default according to the docs, which is the setting I'm using. This seems to be what I need - the sniff may not assume that my array condition is already a boolean, as it could be something else (depending on what `foo` returns); if it _would_ assume that (i.e. be `true`), then the ternary operator would indeed correctly be redundant as `$array['test']` already evaluates to a boolean.

In this case, however, `$array['test']` isn't a boolean if `foo` returns something else:

```php
function foo(): int
{
    return 5;
}

$array = ['test' => foo()];

// $bar is now "true".
$bar = $array['test'] ? true : false;

// $bar is now 5.
$bar = $array['test'];
```

If you don't have type hints or casts in place where `$bar` is later used, you are passing around an integer where you expected a boolean, hence the ternary operator to prevent this problem. In other words, it does serve a purpose here, so it's not really useless as reported.

Although, perhaps I'm misunderstanding the sniff's purpose?

**Comment by kukulich** [comment]
The sniff reports useless ternary operator. The ternary operator in your condition is useless. You should rewrite your condition :)

`assumeAllConditionExpressionsAreAlreadyBoolean` is only for fixer, so it does not produce invalid fixes.

**Comment by tominventisbe** [comment]
I'm afraid I still don't understand :smile: . How is it useless? In the above example, if I leave it out, it breaks code without type hints expecting a boolean instead of an int.

In other words:
* [This is a shortened version of my original code](https://3v4l.org/IIghI).
* [This is the version after rewriting my condition](https://3v4l.org/CN4cV).

As you can see, my program has now effectively changed in behavior, thus the ternary operator may not be left out (without rewriting other code as well), thus it is not useless, thus it is a false positive.

Alternatively, the error message could state something like "_Likely_ useless ternary operator" to clarify that it isn't certain if it is useless. In that case it also shouldn't be autofixed as the fix might break code.

Though I may be missing something here :-).

**Comment by kukulich** [comment]
@tominventisbe 

Eg.
```
$bar = !empty($array['test']);
```
